asmcomp/cmmgen.ml

```diff
@@ -583,31 +580,34 @@ let test_bool dbg cmm =
 let box_float dbg c = Cop(Calloc, [alloc_float_header dbg; c], dbg)
 
 let map_ccatch f rec_flag handlers body =
-  let handlers = List.map
-      (fun (n, ids, handler, dbg) -> (n, ids, f handler, dbg))
-      handlers in
+  let handlers =
+    List.map (fun (n, ids, handler) ->
+        let handler = Cmm.block handler.block_dbg (f handler.expr) in
+        n, ids, handler)
+      handlers
+  in
   Ccatch(rec_flag, handlers, f body)
 
 let rec unbox_float dbg cmm =
   match cmm with
   | Cop(Calloc, [Cblockheader (header, _); c], _) when header = float_header ->
       c
   | Clet(id, exp, body) -> Clet(id, exp, unbox_float dbg body)
-  | Cifthenelse(cond, ifso_dbg, e1, ifnot_dbg, e2, dbg) ->
-      Cifthenelse(cond,
-        ifso_dbg, unbox_float dbg e1,
-        ifnot_dbg, unbox_float dbg e2,
-        dbg)
+  | Cifthenelse(cond, e1, e2, dbg) ->
+      Cifthenelse(cond, unbox_float_block e1, unbox_float_block e2, dbg)
   | Csequence(e1, e2) -> Csequence(e1, unbox_float dbg e2)
   | Cswitch(e, tbl, el, dbg') ->
-    Cswitch(e, tbl,
-      Array.map (fun (expr, dbg) -> unbox_float dbg expr, dbg) el, dbg')
+    Cswitch(e, tbl, Array.map (fun act -> unbox_float_block act) el, dbg')
   | Ccatch(rec_flag, handlers, body) ->
     map_ccatch (unbox_float dbg) rec_flag handlers body
-  | Ctrywith(e1, id, e2, dbg) ->
-      Ctrywith(unbox_float dbg e1, id, unbox_float dbg e2, dbg)
+  | Ctrywith(e1, id, e2) ->
+      Ctrywith(unbox_float dbg e1, id, unbox_float_block e2)
   | c -> Cop(Cload (Double_u, Immutable), [c], dbg)
 
+and unbox_float_block block =
+  let dbg = block.block_dbg in
+  Cmm.block dbg (unbox_float dbg block.expr)
```

Comment:
(I don't know if this was already in the previous patch but this two-layer recursion scheme is actually quite elegant. Well done!)

Reply:
This is new.